src/finalizer/utils/zkSync.ts

```diff
@@ -0,0 +1,229 @@
+  l2MessageIndex: number;
+  l2TxNumberInBlock: number;
+  message: string;
+  sender: string;
+  proof: string[];
+};
+
+const TransactionStatus = zkTypes.TransactionStatus;
+
+/**
+ * @returns Withdrawal finalizaton calldata and metadata.
+ */
+export async function zkSyncFinalizer(
+  logger: winston.Logger,
+  signer: Wallet,
+  hubPoolClient: HubPoolClient,
+  spokePoolClient: SpokePoolClient,
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  _latestBlockToFinalize: number
+): Promise<FinalizerPromise> {
+  const { chainId: l1ChainId } = hubPoolClient;
+  const { chainId: l2ChainId } = spokePoolClient;
+
+  const l1Provider = hubPoolClient.hubPool.provider;
+  const l2Provider = new zksProvider(getNodeUrlList(l2ChainId, 1)[0]);
+  const wallet = new zkWallet(signer.privateKey, l2Provider, l1Provider);
+
+  const tokensBridged = spokePoolClient.getTokensBridged();
+  const { ready, pending } = await sortWithdrawals(l2Provider, tokensBridged);
+  if (pending.length > 0) {
+    logger.debug({
+      at: "zkSyncFinalizer",
+      message: `Found ${pending.length} pending withdrawal transactions.`,
+      pending,
+    });
+  }
+
+  const candidates = await filterMessageLogs(wallet, l2Provider, ready);
+  if (candidates.length > 0) {
+    logger.debug({
+      at: "zkSyncFinalizer",
+      message: `Found ${candidates.length} withdrawals ready to be finalized.`,
+      candidates,
+    });
+  }
+
+  const withdrawalParams = await getWithdrawalParams(wallet, candidates);
+  const txns = await prepareFinalizations(l1ChainId, l2ChainId, withdrawalParams);
+
+  const withdrawals = candidates.map(({ l2TokenAddress, amountToReturn }) => {
+    const l1TokenCounterpart = hubPoolClient.getL1TokenCounterpartAtBlock(
+      l2ChainId,
+      l2TokenAddress,
+      hubPoolClient.latestBlockNumber
+    );
+    const { decimals, symbol: l1TokenSymbol } = hubPoolClient.getTokenInfo(l1ChainId, l1TokenCounterpart);
+    const amountFromWei = convertFromWei(amountToReturn.toString(), decimals);
+    const _withdrawal: Withdrawal = {
+      l2ChainId,
+      l1TokenSymbol,
+      amount: amountFromWei,
+      type: "withdrawal",
+    };
+
+    return _withdrawal;
+  });
+
+  return { callData: txns, withdrawals };
+}
+
+/**
+ * @param provider zkSync L2 provider instance (must be of type zksync-web3.Provider).
+ * @param tokensBridged Array of TokensBridged events to evaluate for finalization.
+ * @returns TokensBridged events sorted according to pending and ready for finalization.
+ */
+async function sortWithdrawals(
+  provider: zksProvider,
+  tokensBridged: TokensBridged[]
+): Promise<{ pending: TokensBridged[]; ready: TokensBridged[] }> {
+  const txnStatus = await Promise.all(
+    tokensBridged.map(({ transactionHash }) => provider.getTransactionStatus(transactionHash))
+  );
+
+  let idx = 0; // @dev Possible to infer the loop index in groupBy ??
+  const { pending = [], ready = [] } = groupBy(tokensBridged, () =>
```

Comment:
Can't this be or am i missing something obvious
```suggestion
  const { pending = [], ready = [] } = groupBy(tokensBridged, (txnStatus) =>
   txnStatus === TransactionStatus.Finalized ? "ready" : "pending"
```